Use es6 classes and arrow functs

juanjoDiaz opened the pull request:
Continuation of #1013

Great to see that stuff is finally going in.

juanjoDiaz commented:
Did anyone had the chance of taking a look at this? 🙂

CendioOssman commented:
Not yet. But hopefully soon. :)

CendioOssman commented on app/localization.js:
```diff
@@ -78,33 +78,32 @@ Localizer.prototype = {
 
             function translateAttribute(elem, attr) {
-                const str = self.get(elem.getAttribute(attr));
+                const str = this.get(elem.getAttribute(attr));
```
These are broken. `this` will be `undefined` in these functions.

CendioOssman commented on core/base64.js:
```diff
@@ -11,8 +11,7 @@ export default {
 
-    encode: function (data) {
-        "use strict";
+    encode(data) {
```
This is a separate cleanup, so could we have that in another commit?

CendioOssman commented on core/display.js:
```diff
@@ -10,109 +10,112 @@
+
+        // ===== EVENT HANDLERS =====
+
+        this.onflush = () => {}; // A flush request has finished
```
I was thinking even further up, so we can avoiding splitting up the two parts of the properties. :)
(this is fine as well though)

juanjoDiaz replied:
I'm not sure if I understand. Where were you planning to put these? Merged with the properties?

CendioOssman replied:
Basically something like this:

```Javascript
        this._tile16x16 = this._drawCtx.createImageData(16, 16);
        Log.Debug("<< Display.constructor");

        // ===== EVENT HANDLERS =====

        this.onflush = () => {}; // A flush request has finished

        // ===== PROPERTIES =====

        this._scale = 1.0;
        this._clipViewport = false;
        this.logo = null;
    }

    // ===== PROPERTIES (cont.) =====

    get scale() { return this._scale; }
    set scale(scale) {
        this._rescale(scale);
    }

```

But those old school event handlers will eventually be removed, so it is perhaps nothing to bother with...

CendioOssman commented on app/ui.js:
```diff
@@ -552,8 +552,7 @@ const UI = {
             bar.style.transitionDuration = '0s';
-            bar.addEventListener('transitionend', function () {
-                this.style.transitionDuration = ""; });
+            bar.addEventListener('transitionend', () => this.style.transitionDuration = '');
```
I don't think this will work as `this` will not be properly rebound to `bar`.

CendioOssman commented on app/webutil.js:
```diff
@@ -119,7 +119,7 @@ export function initSettings (callback /*, ...callbackArgs */) {
+        window.chrome.storage.sync.get((cfg) => {
             settings = cfg;
             if (callback) {
                 callback.apply(this, callbackArgs);
```
`this` will change value here. Have you checked the callers and made sure they are unaffected?

juanjoDiaz replied:
Yes. The context is never used in the callbacks use through the application

This change also makes the scope of the callback consistent regardless of whether chrome storage is used or not. Before it was different.

CendioOssman commented on karma.conf.js:
```diff
@@ -99,6 +99,7 @@ module.exports = function(config) {
       'tests/test.*.js': ['babel'],
       'tests/fake.*.js': ['babel'],
       'tests/assertions.js': ['babel'],
+      'tests/payback*.js': ['babel'],
```
Not part of tests, so should not be included.

CendioOssman commented on tests/test.rfb.js:
```diff
@@ -23,18 +23,18 @@ import sinon from '../vendor/sinon.js';
 })();
 
-const push8 = function (arr, num) {
+const push8 = (arr, num) => {
```
Wouldn't it be better to turn these in to normal functions instead?

juanjoDiaz replied:
What do you mean?

`push8(arr, num)` // Not possible since it's not in a function
`const push8 = function (arr, num) {` Zero advantages since it´s a pure function.
`function push8(arr, num) {` I prefer this notation but I tried to minimize the changes.

CendioOssman replied:
The last one. You already had a bunch of such transformations, so we've already started down that path.

CendioOssman commented on utils/genkeysymdef.js:
```diff
@@ -108,7 +108,7 @@ out +=
 "\n" +
 "export default {\n" +
-"    lookup : function(u) {\n" +
+"    lookup(u) {\n" +
```
Group this with other other function declaration cleanups in to a separate commit.

juanjoDiaz replied:
What do you mean?
There are 2 commits:
1.- Change the objects to classes
2.- Change all the functions to arrow functions

(this change is part of the second one)

CendioOssman replied:
I'd like a third one:

3.- Consistent function declarations (`foo: function()` => `function foo()`, `const foo = function()` => `function foo()`, etc.)

juanjoDiaz replied:
I see. Currently:
1.- The shorthand notation is used if the function is a property of an object or class `{ foo() { ... } }` or `class bar { foo() { ... } }`
2.- The fat arrow notation `const foo = () => { ... };` for callback and any other function declared.
3.- The calssic `function foo() { ... }` for top level functions or functions that depend on the scope.

I prefer `const foo = () => { ... };` over `function foo() { ... }` for callbacks and functions that are not declared at the top level since it ensures that `foo` can't be overriden and automatically binds the function scope.

Would you like me to do a 3rd commit replacing all the `const foo = () => { ... };` by `function foo() { ... }` regardless of where they are?

CendioOssman replied:
> So following this logic, lookup is correct as a shorthand function. It wouldn't even be a classic function unless we move it out of the object.

Fair enough. Might want to note that in the commit message in that case so that people aren't confused that the commit modifies more than just classes.

> I generally prefer const foo = () => { ... }; over function foo() { ... } for callbacks and functions that are not declared at the top level since it ensures that foo can't be overriden and automatically binds the function scope.
> 

What do you mean when you say it binds the function scope? Is there some practical detail I'm overlooking here?

I'm just concerned that `const foo = ()` is very similar to the `foo: function()` style we want to move away from. Also, coming from other languages, `const foo` reads more like "reference to an existing function" than a function definition, so that is tripping me up a bit.

juanjoDiaz replied:
I meant that
`const foo = () => { ... };`
is equivalent to
```
function foo() { ... }
foo.bind(this)
```
so you don't need to do stuff like 
```
const self = this;
function foo() { ... <use self> ... }
```

Other difference is that const is safer to accidental overriding.
```
const foo = () => { ... };
foo = 2; // Throws an error saying that you can't override foo
foo();
```
while 
```
function foo() { ... }
foo = 2;
foo(); // Throws an error because foo is not a function
```

So nothing major. I think that my 3 points above summarize well my point of view :)

CendioOssman commented on utils/use_require.js:
```diff
@@ -39,15 +39,14 @@ const no_transform_files = new Set([
         const args = Array.prototype.slice.call(arguments);
         return new Promise((resolve, reject) => {
-            original.apply(obj, args.concat((err, value) => {
+            original.apply(this, args.concat((err, value) => {
```
This will be incorrect. `this` needs to be from the call to the unnamed function. With an arrow function you'll instead get `this` from the `promisify()` call.

CendioOssman commented on utils/use_require.js:
```diff
@@ -70,29 +69,26 @@ const babelTransformFile = promisify(babel.transformFile);
-            });
-        }));
+        return Promise.all(paths.map(filepath => lstat(filepath) .then((stats) => {
+            if (filter !== undefined && !filter(filepath, stats)) return;
```
It is usually good form to keep every `.then` on a new line for readability.

CendioOssman commented on utils/use_require_helpers.js:
```diff
@@ -4,11 +4,10 @@ const path = require('path');
         const args = Array.prototype.slice.call(arguments);
         return new Promise((resolve, reject) => {
-            original.apply(obj, args.concat((err, value) => {
+            original.apply(this, args.concat((err, value) => {
```
See above.

juanjoDiaz commented:
I think that everything is fixed except for the couple question that I've made above.

The failing CI seems to be one of those weird Travis issues unrelated to the code.

CendioOssman commented on .eslintrc:
```diff
@@ -11,6 +11,10 @@
         "no-var": "error",
-        "no-useless-constructor": "error"
+        "no-useless-constructor": "error",
+        "prefer-arrow-callback": ["error", { "allowNamedFunctions": false, "allowUnboundThis": true }],
```
These are the default options. Can't we just omit them in that case?

juanjoDiaz replied:
I guess that we could remove the options yeah.
I just copy-pasted them from air-bnb style config
I guess that it's more explicit and less prone to errors when updating in case eslint changes defaults.

But I can remove them if you prefer that :)

CendioOssman replied:
Very subjective I guess, but personally I would read such things as deviations from the defaults. :)

CendioOssman commented on .eslintrc:
```diff
@@ -11,6 +11,10 @@
+        "prefer-arrow-callback": ["error", { "allowNamedFunctions": false, "allowUnboundThis": true }],
+        "arrow-body-style": ["error", "as-needed", { "requireReturnForObjectLiteral": false }],
+        "arrow-parens": ["error", "as-needed", { "requireForBlockBody": true }],
+        "arrow-spacing": ["error", { "before": true, "after": true }], "no-confusing-arrow": ["error", { "allowParens": true }],
```
Need a newline here for the last rule

juanjoDiaz replied:
Do you mean that the line is too long?

CendioOssman commented on utils/use_require.js:
```diff
@@ -39,15 +39,14 @@ const no_transform_files = new Set([
 function promisify(original) {
-    return function () {
-        const obj = this;
+        return function () {
```
The indentation is off here.

CendioOssman commented:
> The failing CI seems to be one of those weird Travis issues unrelated to the code.

Looks like it, yes. You can usually restart the job in those cases and get it to pass.

juanjoDiaz commented:
I've removed the defaults from the eslint config, added a third commit normalizing the function declarations and added the rationale for that in the commit message as you suggested.

Is there anything else left?

CendioOssman commented:
Almost. Some details were a bit off in the last commit:

a) The rationale should be split out to the relevant commit. The three parts seem to correspond to the three commits, unless I'm missing something?

b) Most of the changes in commit 3 touch the same functions as in the earlier two commits. Let's avoid the initial steps and just change them once.

c) The final result of the encoding helper functions is broken as we now got an incorrect binding of `this`.

juanjoDiaz commented:
Alright.

a) Removed the unneeded changes from the first to commits that were overridden by the third commit
b) Fixed the commit message
c) Can you clarify? Do you mean the `encode` method in `Base64`? That shouldn't be broken since it was an old style function and not it's a shorthand function which is not automatically bound like fat arrow functions. Or you were talking about something else?

CendioOssman commented:
It's part of the Tight decoder. E.g. `indexedToRGBX2Color()`.

samhed commented on tests/test.webutil.js:
```diff
@@ -117,9 +117,9 @@ describe('WebUtil', function() {
-                            remove: function() {}
+                            get(cb){ cb(settings); },
+                            set(){},
+                            remove() {}
```
Inconsistent spacing between () and {}. I prefer a space like you did for remove() {}.

samhed commented on core/input/keyboard.js:
```diff
@@ -307,7 +307,7 @@ export default class Keyboard {
         const target = this._target;
         const downList = this._keyDownList;
-        ['AltLeft', 'AltRight'].forEach(function (code) {
+        ['AltLeft', 'AltRight'].forEach((code) => {
```
In the other places you removed the parentheses around the parameter:

    code => { ... }

However, I think I prefer leaving the parentheses in like this, but it should at least be consistent.

samhed commented on core/rfb.js:
```diff
@@ -199,17 +199,17 @@ export default class RFB extends EventTargetMixin {
-        }.bind(this));
-        this._sock.on('close', function (e) {
+        });
+        this._sock.on('close', (e) => {
```
another inconsistency with (e) => { ... } vs e => { ... }

samhed commented on core/rfb.js:
```diff
@@ -454,18 +452,15 @@ export default class RFB extends EventTargetMixin {
 
         Log.Info("Encoding stats for this connection:");
-        Object.keys(stats).forEach(function (key) {
+        Object.keys(stats).forEach((key) => {
```
dito

samhed commented on core/util/events.js:
```diff
@@ -120,7 +120,7 @@ export function releaseCapture () {
         // There might be events already queued, so we need to wait for
         // them to flush. E.g. contextmenu in Microsoft Edge
-        window.setTimeout(function(expected) {
+        window.setTimeout((expected) => {
```
inconsistency

samhed commented on core/util/eventtarget.js:
```diff
@@ -32,7 +32,7 @@ export default class EventTargetMixin {
          return true;
       }
-      this._listeners.get(event.type).forEach(function (callback) {
+      this._listeners.get(event.type).forEach((callback) => {
```
dito

samhed commented on core/util/logging.js:
```diff
@@ -24,7 +24,7 @@ export function init_logging (level) {
     }
 
-    Debug = Info = Warn = Error = function (msg) {};
+    Debug = Info = Warn = Error = () => {};
```
did you mean to remove the 'msg' parameter?

samhed commented on core/websock.js:
```diff
@@ -225,7 +223,7 @@ export default class Websock {
             }
 
-            this._websocket.onmessage = function (e) { return; };
+            this._websocket.onmessage = () => {};
```
'e' went missing?

samhed commented on core/websock.js:
```diff
@@ -196,25 +194,25 @@ export default class Websock {
-        }).bind(this);
-        this._websocket.onclose = (function (e) {
+        };
+        this._websocket.onclose = (e) => {
```
inconsistency

samhed commented on core/websock.js:
```diff
@@ -196,25 +194,25 @@ export default class Websock {
-        }).bind(this);
-        this._websocket.onerror = (function (e) {
+        };
+        this._websocket.onerror = (e) => {
```
inconsistency

samhed commented on karma.conf.js:
```diff
@@ -1,6 +1,6 @@
 // Karma configuration
 
-module.exports = function(config) {
+module.exports = (config) => {
```
inconsistency

samhed commented on po/xgettext-html:
```diff
@@ -89,7 +89,7 @@ for (let i = 0; i < opt.argv.length; i++) {
     const body = dom.window.document.body;
 
-    const locator = function (elem) {
+    const locator = (elem) => {
```
dito

samhed commented on tests/playback.js:
```diff
@@ -96,9 +96,9 @@ export default class RecordingPlayer {
-        this._rfb._sock.send = function (arr) {};
-        this._rfb._sock.close = function () {};
-        this._rfb._sock.flush = function () {};
+        this._rfb._sock.send = () => {};
```
'arr' went missing?

CendioOssman commented:
The parenthesis were discussed here https://github.com/novnc/noVNC/pull/1013#discussion-diff-186998484L74 (open in private mode and it should load)

juanjoDiaz commented:
Fix the issue with the rfb functions scope.

Following my own definition, those are functions that should be bound and are passed around so they are good as `const foo = (...) => { ... }`


Regarding @samhed comment. I have a nother PR fixing all the spacing inconsistencies (currently there are 11013 in the code 🙂). It was part of #1013, the mega-PR. Since now I'm  slowly pouring those in small PRs, I'll just create a next PR fixing all those once this one gets in if that's ok.

samhed commented:
ok

juanjoDiaz commented:
I just rebased on top of master and fix the conflicts once again.

@CendioOssman anything else pending?

CendioOssman commented:
I don't think so. So let's merge this. Thanks for your hard work. :)

CendioOssman commented:
Btw, you don't generally need to rebase to fix conflicts unless there is something major. We can sort out the details in most cases when we merge.

CendioOssman commented:
Hang on, something is broken with IE. Need to investigate.

CendioOssman commented:
False alarm (sort-of). We have an existing issue where if you run `npm test` by hand you won't get the transformations necessary to run things in IE.
